### PR TITLE
Updated The Navigator To Use New Config Object System

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Start the migration to passing configurations via dependency injection instead of just as global variables. The configuration is now grabbed from the global variable in `full_system_main` and passed through the navigator and the components below it. As such, all the configs in the navigator *should* now be dependency injected. 

Note that unit tests still grab the global config instead of their own config, so they are still sensitive to default parameter changes in the yaml.

Had to perform some restructuring in the Navigator to make the config structure, most notably by creating an `ObstacleFactory` for creating obstacles instead of just having free-floating functions.

One thing I'm really no a fan of is the significant nesting in the initializer lists this causes, most notably in `ai.cpp`. Would really appreciate if folks could take a look and think of a slightly more elegant way.

### Testing Done

Unit tests.

### Resolved Issues
Part of https://github.com/UBC-Thunderbots/Software/issues/960

### Length Justification

Configuration changes are fairly wide-reaching, many changes are just renames. Also I created a AIConfig in `ai.yaml` that resulted in a lot of re-indenting.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Seems good overall. One minor question.

File: src/software/ai/navigator/path_manager/velocity_obstacle_path_manager.cpp [hunk 1]
```diff
@@ -43,12 +38,17 @@ const std::map<RobotId, std::optional<Path>> VelocityObstaclePathManager::getMan
         // store velocity obstacle for current path
         if (path && path->size() >= 2)
         {
+            // We want to avoid the start of every other path, assuming that
+            // there is a robot moving along the path from the path's start
             std::vector<Point> path_points = path->getKnots();
+            Vector initial_path_velocity =
+                (path_points[1] - current_objective.start)
+                    .normalize(current_objective.current_speed);
+            Robot mock_path_robot(0, current_objective.start, initial_path_velocity,
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/navigator/path_manager/velocity_obstacle_path_manager.cpp:47]: Isn't it embedded in `mock_path_robot`?

[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_manager/velocity_obstacle_path_manager.cpp:47]: yeah this looks ok to me

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Overall looks good. Just few minor comments

File: src/software/ai/ai.h [hunk 2]
```diff
@@ -14,10 +14,13 @@
...
    public:
+    AI() = delete;
+
     /**
-     * Creates a new AI
+     * Create an AI with given configuration
+     * @param config The AI configuration
      */
-    explicit AI();
+    AI(std::shared_ptr<const AIConfig> config);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/ai.h]: changed

File: src/software/ai/ai_wrapper.h [hunk 3]
```diff
@@ -18,7 +18,14 @@ class AIWrapper : public ThreadedObserver<World>,
...
    public:
-    AIWrapper() = default;
+    AIWrapper() = delete;
+
+    /**
+     * Create an AI with the given config
+     *
+     * @param config The AI configuration
+     */
+    AIWrapper(std::shared_ptr<const AIConfig> config);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/ai_wrapper.h]: changed

File: src/software/ai/ai_wrapper.cpp [hunk 4]
```diff
@@ -13,7 +15,7 @@ void AIWrapper::onValueReceived(World world)
 
 void AIWrapper::runAIAndSendPrimitives()
 {
-    if (Util::DynamicParameters->getAIConfig()->RunAI()->value())
+    if (Util::DynamicParameters->getAIControlConfig()->RunAI()->value())
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/ai_wrapper.cpp:18]: This uses the `AIControl` config, which is different from the `AIConfig`. I agree we should use that here, but I think it's out-of-scope for this PR.

File: src/software/ai/navigator/path_manager/velocity_obstacle_path_manager.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/navigator/path_manager/velocity_obstacle_path_manager.cpp:47]: Oh I see we're using the velocity to embed that data. All good

**Review by MathewMacDougall — COMMENTED** [review]
No other comments from me, just need to CI to pass

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI!

**Review by jonathanlew — APPROVED** [review]
Re-approving...

**Review by MathewMacDougall — APPROVED** [review]

**Pull request merged**